### PR TITLE
cleanup: move scheduler unit tests to use PodWrapper

### PR DESCRIPTION
**Pull request by yibozhuang**
#### What type of PR is this?

/kind cleanup

#### What this PR does / why we need it:

Move to use testing `PodWrapper` where applicable to
reduce duplicating pod creation code and shorten
number of lines.

Adding additional wrapper functions in PodWrapper
to ensure it covers different pod spec under different plugins
and test cases.

Signed-off-by: Yibo Zhuang <yibzhuang@gmail.com>

#### Special notes for your reviewer:

This is a follow PR and builds on top of #109536 

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

**Review by yuanchen8911 — APPROVED** [review]
Thanks for the PR. /lgtm

**Comment by Huang-Wei** [comment]
Thanks for the PR. Please fix the CI failures and I will review.

File: pkg/scheduler/testing/wrappers.go
Comment (kerthcet) [inline pkg/scheduler/testing/wrappers.go:277]: `p.Status.Conditions = append(p.Status.Conditions, conditions...)`

File: pkg/scheduler/testing/wrappers.go
Comment (kerthcet) [inline pkg/scheduler/testing/wrappers.go:307]: Why not the all Containers?

File: pkg/scheduler/testing/wrappers.go
Comment (kerthcet) [inline pkg/scheduler/testing/wrappers.go:458]: CreationTimestamp?

**Comment by kerthcet** [comment]
thanks @yibozhuang for your wonderful work, and I left some comments. Maybe @Huang-Wei will take a further review after.

File: pkg/scheduler/framework/plugins/interpodaffinity/filtering_test.go
Comment (sanposhiho) [inline pkg/scheduler/framework/plugins/interpodaffinity/filtering_test.go:55]: You can also use pod wrapper in `createPodWithAffinityTerms`.

File: pkg/scheduler/internal/queue/scheduling_queue_test.go
Comment (sanposhiho) [inline pkg/scheduler/internal/queue/scheduling_queue_test.go:64]: Please use PodWrapper as well for medPriorityPodInfo

File: pkg/scheduler/testing/wrappers.go
Comment (sanposhiho) [inline pkg/scheduler/testing/wrappers.go:450]: Given `Label` method is exist, we should name this as `Labels`?

File: pkg/scheduler/internal/queue/scheduling_queue_test.go
Comment (yibozhuang) [reply pkg/scheduler/internal/queue/scheduling_queue_test.go:64]: `medPriorityPodInfo` and `unschedulablePodInfo` have annotations which requires #109536 to be merged first since that adds support for Pod annotations so I think that can be a separate PR or I can rebase after.

File: pkg/scheduler/testing/wrappers.go
Comment (yibozhuang) [reply pkg/scheduler/testing/wrappers.go:450]: `Labels` is a Pod attribute so cannot reuse that as a function or it will conflict.

File: pkg/scheduler/testing/wrappers.go
Comment (yibozhuang) [reply pkg/scheduler/testing/wrappers.go:458]: `CreationTimestamp ` is a Pod attribute so cannot reuse that as a function or it will conflict.

File: pkg/scheduler/extender_test.go
Comment (yibozhuang) [reply pkg/scheduler/extender_test.go:369]: fixed.

File: pkg/scheduler/extender_test.go
Comment (yibozhuang) [reply pkg/scheduler/extender_test.go:361]: Thanks, changed to "0".

File: pkg/scheduler/framework/plugins/interpodaffinity/filtering_test.go
Comment (yibozhuang) [reply pkg/scheduler/framework/plugins/interpodaffinity/filtering_test.go]: Sure thing, done.

File: pkg/scheduler/framework/plugins/interpodaffinity/filtering_test.go
Comment (yibozhuang) [reply pkg/scheduler/framework/plugins/interpodaffinity/filtering_test.go]: Thanks, done.

File: pkg/scheduler/framework/plugins/noderesources/balanced_allocation_test.go
Comment (yibozhuang) [reply pkg/scheduler/framework/plugins/noderesources/balanced_allocation_test.go]: Fixed.

File: pkg/scheduler/framework/plugins/noderesources/balanced_allocation_test.go
Comment (yibozhuang) [reply pkg/scheduler/framework/plugins/noderesources/balanced_allocation_test.go]: Fixed.

File: pkg/scheduler/framework/plugins/noderesources/requested_to_capacity_ratio_test.go
Comment (yibozhuang) [reply pkg/scheduler/framework/plugins/noderesources/requested_to_capacity_ratio_test.go:351]: The current code uses `extendedResourcePod1` since these are pod resources so I think keeping the name as it is makes more sense.

File: pkg/scheduler/framework/plugins/noderesources/requested_to_capacity_ratio_test.go
Comment (yibozhuang) [reply pkg/scheduler/framework/plugins/noderesources/requested_to_capacity_ratio_test.go:355]: Same, using the current name makes more sense since it is pod resources.

File: pkg/scheduler/framework/plugins/nodevolumelimits/non_csi_test.go
Comment (yibozhuang) [reply pkg/scheduler/framework/plugins/nodevolumelimits/non_csi_test.go]: Sure thing, did the refactor to move all the pod using PVCs to the top and have all tests consume them.

File: pkg/scheduler/framework/plugins/volumerestrictions/volume_restrictions_test.go
Comment (yibozhuang) [reply pkg/scheduler/framework/plugins/volumerestrictions/volume_restrictions_test.go:254]: Done.

File: pkg/scheduler/framework/plugins/volumerestrictions/volume_restrictions_test.go
Comment (yibozhuang) [reply pkg/scheduler/framework/plugins/volumerestrictions/volume_restrictions_test.go:273]: Done.

File: pkg/scheduler/testing/wrappers.go
Comment (yibozhuang) [reply pkg/scheduler/testing/wrappers.go]: Thanks, fixed.

File: pkg/scheduler/testing/wrappers.go
Comment (yibozhuang) [reply pkg/scheduler/testing/wrappers.go]: Changed to allow a container to be created and then assign the ports to the new container, thanks for catching this.

File: pkg/scheduler/framework/plugins/interpodaffinity/filtering_test.go
Comment (yibozhuang) [reply pkg/scheduler/framework/plugins/interpodaffinity/filtering_test.go:55]: Thanks, converted `createPodWithAffinityTerms` to use `PodWrapper`

File: pkg/scheduler/testing/wrappers.go
Comment (Huang-Wei) [inline pkg/scheduler/testing/wrappers.go:310]: Not sure I like it as `index` may be out of bound.

Is there any case `ContainerPort` cannot fit?

File: pkg/scheduler/testing/wrappers.go
Comment (Huang-Wei) [inline pkg/scheduler/testing/wrappers.go:554]: Prefer `InitReq`

**Review by Huang-Wei — COMMENTED** [review]
Thanks Yibo. Some comments below.

File: pkg/scheduler/internal/queue/scheduling_queue_test.go
Comment (sanposhiho) [reply pkg/scheduler/internal/queue/scheduling_queue_test.go:64]: But you've already implemented `CreationTimestamp` in this PR which is included in https://github.com/kubernetes/kubernetes/pull/109536, right?
https://github.com/kubernetes/kubernetes/pull/109536/files#diff-b8c43b6841e30bdec0f6bc6436e7d4c1bb3a67073dfc7efafcc4522429458961R214
So, I think you need to rebase after #109536 merged regardless of this change.

File: pkg/scheduler/testing/wrappers.go
Comment (sanposhiho) [reply pkg/scheduler/testing/wrappers.go:519]: Really? I think we can because we already have `Name` method on PodWrapper.

Comment (Huang-Wei) [reply pkg/scheduler/testing/wrappers.go:519]: > Labels is a Pod attribute so cannot reuse that as a function or it will conflict.

We can actually :) In that case we should specify the full path rather than emitting the anonymous field (ObjectMeta):

```go
// Label sets a {k,v} pair to the inner pod.
func (p *PodWrapper) Label(k, v string) *PodWrapper {
	if p.ObjectMeta.Labels == nil {
		p.ObjectMeta.Labels = make(map[string]string)
	}
	p.ObjectMeta.Labels[k] = v
	return p
}

// Labels sets all {k,v} pair provided by `labels` to the inner pod.
func (p *PodWrapper) Labels(labels map[string]string) *PodWrapper {
	for k, v := range labels {
		p.Label(k, v)
	}
	return p
}
```

File: pkg/scheduler/framework/plugins/noderesources/requested_to_capacity_ratio_test.go
Comment (kerthcet) [reply pkg/scheduler/framework/plugins/noderesources/requested_to_capacity_ratio_test.go:351]: But it's a bad naming I think. `extendedResourcePod1` indicates it's a pod not the pod resource. We used to name this   `extendedResourcePod1` because it's actually a pod.

File: pkg/scheduler/testing/wrappers.go
Comment (kerthcet) [reply pkg/scheduler/testing/wrappers.go:527]: `p.ObjectMeta.CreationTimestamp = t`  works.

Comment (yibozhuang) [reply pkg/scheduler/testing/wrappers.go]: Yes you are right, sorry, fixed now.

File: pkg/scheduler/internal/queue/scheduling_queue_test.go
Comment (yibozhuang) [reply pkg/scheduler/internal/queue/scheduling_queue_test.go]: Yea that makes sense. I've updated to use PodWrapper for `medPriorityPodInfo` and `unschedulablePodInfo` and will rebase.

File: pkg/scheduler/testing/wrappers.go
Comment (yibozhuang) [reply pkg/scheduler/testing/wrappers.go]: Got it, thanks!
Fixed.

File: pkg/scheduler/testing/wrappers.go
Comment (yibozhuang) [reply pkg/scheduler/testing/wrappers.go]: Thanks, fixed.

File: pkg/scheduler/testing/wrappers.go
Comment (yibozhuang) [reply pkg/scheduler/testing/wrappers.go]: Fixed.

File: pkg/scheduler/testing/wrappers.go
Comment (yibozhuang) [reply pkg/scheduler/testing/wrappers.go]: The only case where `ContainerPort` doesn't fit this `MakePod` at https://github.com/kubernetes/kubernetes/blob/master/pkg/scheduler/internal/cache/cache_test.go#L1739

Since when we call `Req()` under `PodWrapper` it creates the container and we want to set the port for that container rather than create and append.

But I am doing something like this for now to have it fit using `ContainerPort`: https://github.com/kubernetes/kubernetes/pull/109684/files#diff-c93487173ed1e8915cdcdd46130ed723d517e191c35e817cfd3f76c913b94281R1707-R1711

File: pkg/scheduler/framework/plugins/noderesources/requested_to_capacity_ratio_test.go
Comment (yibozhuang) [reply pkg/scheduler/framework/plugins/noderesources/requested_to_capacity_ratio_test.go:350]: Sure, updated.

File: pkg/scheduler/framework/plugins/noderesources/requested_to_capacity_ratio_test.go
Comment (yibozhuang) [reply pkg/scheduler/framework/plugins/noderesources/requested_to_capacity_ratio_test.go:354]: Updated.

File: pkg/scheduler/testing/wrappers.go
Comment (Huang-Wei) [reply pkg/scheduler/testing/wrappers.go]: I feel like a ContainerWrapper is worthy, similar to NodeSelectorWrapper as a "second-level" wrapper, which is passed PodWrapper:

```go
st.MakePod().Containers(st.MakeContainer().Req().Port().Obj(), ...)
```

Comment (yibozhuang) [reply pkg/scheduler/testing/wrappers.go]: Sure thing, created `ContainerWrapper` and refactored the existing `PodWrapper` code to leverage this. I agree that this does look cleaner.

**Review by sanposhiho — APPROVED** [review]
Looks good to me. 
Let's`/hold` this until https://github.com/kubernetes/kubernetes/pull/109536 is merged.

/lgtm 
/hold

**Review by ahg-g — COMMENTED** [review]
This is a great cleanup, thanks for doing it.

This is hard to review in detail, but can you confirm that no test in this PR changed semantically?

File: pkg/scheduler/extender_test.go
Comment (Huang-Wei) [reply pkg/scheduler/extender_test.go:393]: we can make MakePod() accept variadic parameters so as to avoid significant changes.

> it seems that in many places we create pods without names

yup, UTs don't really need a pod name.

File: pkg/scheduler/framework/plugins/noderesources/balanced_allocation_test.go
Comment (yibozhuang) [reply pkg/scheduler/framework/plugins/noderesources/balanced_allocation_test.go]: sure thing, renamed all `machine` naming to `node`

File: pkg/scheduler/framework/plugins/helper/spread_test.go
Comment (yibozhuang) [reply pkg/scheduler/framework/plugins/helper/spread_test.go]: sure thing, done.

**Comment by yibozhuang** [comment]
> This is hard to review in detail, but can you confirm that no test in this PR changed semantically?

Yea I confirmed as I was doing the clean up that the pod spec after the change matches with the previous pod spec.

File: pkg/scheduler/extender_test.go
Comment (yibozhuang) [reply pkg/scheduler/extender_test.go:393]: > we can make MakePod() accept variadic parameters so as to avoid significant changes.
What would the `MakePod()` look like for this case?

Are you imagining something like?

```go
func MakePod(name ...string) *PodWrapper {
	wrapper := &PodWrapper{v1.Pod{}}
	if len(name) >= 1 {
                wrapper.SetName(name[0])
        }
        return wrapper
}
```

This will be kind of weird if the caller is specifying multiple values as only the first will be used.

> yup, UTs don't really need a pod name.
I agree here that UTs don't necessarily need specify a name so enforcing may not be ideal.

Maybe we can consider doing a separate clean up for this case in a different PR?

Comment (Huang-Wei) [reply pkg/scheduler/extender_test.go:393]: I'm fine with leaving it as a follow-up.

Comment (ahg-g) [reply pkg/scheduler/extender_test.go:393]: sgtm

**Pull request merged**
